src/services/summary/hooks/trigger-calculation.js

```diff
@@ -0,0 +1,88 @@
+const _ = require('lodash');
+const moment = require('moment');
+const ActivityLog = require('../../activity-log/activity-log-model');
+const actionConstants = require('../../../constants/actions');
+const Summary = require('../../summary/summary-model');
+
+// triggerCalculation is a before hook function to trigger a summary
+// calculation on specified dates and organization.
+// If no date specifies, it will calculate today's summary.
+const triggerCalculation = () => (hook) => {
+  const startDate = hook.params.query.start_date;
+  const endDate = hook.params.query.end_date;
+  const organizationName = hook.params.query.organization;
+  const trigger = hook.params.query.trigger;
+  if (!trigger) {
+    return Promise.resolve(hook);
+  }
+  let timestamp = {};
+  // Change start_date and end_date to mongoose query format.
+  if (startDate) {
+    timestamp.$gte = startDate;
+  }
+  if (endDate) {
+    timestamp.$lte = endDate;
+  }
+  if (_.isEmpty(timestamp)) {
+    timestamp = moment().format('YYYY-MM-DD');
+  }
+  // Start a calculation
+  const query = {
+    organization: organizationName,
+    timestamp,
+  };
+  // Remap state to summary's field name.
+  const remapStateToFieldName = {};
+  remapStateToFieldName[actionConstants.ASSIGN] = 'assigned';
+  remapStateToFieldName[actionConstants.PROCESS] = 'processing';
+  remapStateToFieldName[actionConstants.RESOLVE] = 'resolved';
+  return ActivityLog.find(query)
+    .then(logs => {
+      // Loop through activity log and summarize each day into table.
+      const summaryTable = {};
+      for (let i = 0, len = logs.length; i < len; ++i) {
+        const date = moment(logs[i].timestamp).format('YYYY-MM-DD');
+        if (!(date in summaryTable)) {
+          summaryTable[date] = {};
+        }
+        const department = logs[i].department;
+        if (!(department in summaryTable[date])) {
+          summaryTable[date][department] = {};
+        }
+        const action = remapStateToFieldName[logs[i].action];
+        // +1 if it is one of [assigned, processing, resolved].
+        if (action) {
+          if (!(action in summaryTable[date][department])) {
+            summaryTable[date][department][action] = 0;
+          }
+          summaryTable[date][department][action]++;
+        }
+      }
+      // Reformat to legitimate input for Restful /summaries service.
+      const allDateSummaries = [];
+      for (const date of Object.keys(summaryTable)) {
+        const oneDaySummary =
+          _.reduce(summaryTable[date], (result, departmentSummary, departmentName) => {
+            result.push(Object.assign({ department: departmentName }, departmentSummary));
+            return result;
+          }, []);
+        allDateSummaries.push({
+          date,
+          organization: organizationName,
+          by_department: oneDaySummary,
```

Comment:
what is the best way to define dictionary schema in Mongoose? Using schema.types.mixed?